File: ios/Classes/classifiers/NfbBridge.swift
import Flutter
import UIKit

// MARK: - NfbBridge

class NfbBridge: NSObject {

    // MARK: Static state (used by C callbacks which cannot capture `self`)

    /// The currently active bridge. C callbacks reach the instance through this
    /// static weak reference because `clCNFB_Set*Event` callbacks have no
    /// `void* context` parameter.
    private static weak var activeBridge: NfbBridge?

    // MARK: Stream handlers

    let nfbStateHandler = DeviceStreamHandler(channelId: "neiry_kit/events/nfbState")
    let nfbErrorHandler = DeviceStreamHandler(channelId: "neiry_kit/events/nfbError")

    /// Returns all (channelId, handler) pairs for EventChannel registration in the plugin.
    func allStreamHandlers() -> [(String, FlutterStreamHandler)] {
        return [
            (nfbStateHandler.channelId, nfbStateHandler),
            (nfbErrorHandler.channelId, nfbErrorHandler),
        ]
    }

    // MARK: Instance state

    private var nfb: OpaquePointer?

    // MARK: - Public API

    /// Creates an uncalibrated NFB classifier for the given device handle.
    func create(device: OpaquePointer) throws {
        if nfb != nil { unregisterCallbacks() }
        var error = clCError()
        nfb = clCNFB_Create(device, &error)
        try checkCError(error)
        registerCallbacks()
    }

    /// Creates a calibrated NFB classifier for the given device handle.
    /// If `calibrationData` is provided, it is imported into the calibrator before creating.
    func createCalibrated(device: OpaquePointer, calibrationData: [String: Any]?) throws {
        if nfb != nil { unregisterCallbacks() }
        guard let calibrator = clCNFBCalibrator_CreateOrGet(device) else {
            throw FlutterError(code: "NULL_HANDLE", message: "clCNFBCalibrator_CreateOrGet returned nil", details: nil)
        }
        if let calibrationData = calibrationData {
            var data = clCIndividualNFBData()
            if let ts = calibrationData["ts"] as? Int {
                data.timestampMilli = Int64(ts)
            }
            if let failReasonRaw = calibrationData["failReason"] as? Int {
                data.failReason = clCIndividualNFBCalibrationFailReason(rawValue: UInt32(failReasonRaw))
            }
            if let v = calibrationData["individualFrequency"] as? Double { data.individualFrequency = Float(v) }
            if let v = calibrationData["individualPeakFrequency"] as? Double { data.individualPeakFrequency = Float(v) }
            if let v = calibrationData["individualPeakFrequencyPower"] as? Double { data.individualPeakFrequencyPower = Float(v) }
            if let v = calibrationData["individualPeakFrequencySuppression"] as? Double { data.individualPeakFrequencySuppression = Float(v) }
            if let v = calibrationData["individualBandwidth"] as? Double { data.individualBandwidth = Float(v) }
            if let v = calibrationData["individualNormalizedPower"] as? Double { data.individualNormalizedPower = Float(v) }
            if let v = calibrationData["lowerFrequency"] as? Double { data.lowerFrequency = Float(v) }
            if let v = calibrationData["upperFrequency"] as? Double { data.upperFrequency = Float(v) }
            var importError = clCError()
            clCNFBCalibrator_ImportIndividualNFBData(calibrator, &data, &importError)
            try checkCError(importError)
        }
        var error = clCError()
        nfb = clCNFB_CreateCalibrated(device, calibrator, &error)
        try checkCError(error)
        registerCallbacks()
    }

    /// Unregisters all callbacks and releases the NFB handle.
    /// There is no `clCNFB_Destroy` in the SDK — disposal is callback-only.
    func dispose() {
        unregisterCallbacks()
        nfb = nil
    }

    // MARK: - Callbacks

    private func registerCallbacks() {
        guard let nfb = nfb else { return }
        NfbBridge.activeBridge = self

        clCNFB_SetOnUserStateChangedEvent(nfb) { _, data in
            guard let bridge = NfbBridge.activeBridge,
                  let data = data else { return }
            let state = data.pointee
            let map: [String: Any] = [
                "ts":    state.timestampMilli,
                "delta": state.delta,
                "theta": state.theta,
                "alpha": state.alpha,
                "smr":   state.smr,
                "beta":  state.beta,
            ]
            bridge.nfbStateHandler.send(map)
        }

        clCNFB_SetOnErrorEvent(nfb) { _, msg in
            guard let bridge = NfbBridge.activeBridge else { return }
            let message = msg.map { String(cString: $0) } ?? ""
            bridge.nfbErrorHandler.send(["message": message])
        }
    }

    private func unregisterCallbacks() {
        guard let nfb = nfb else { return }
        clCNFB_SetOnUserStateChangedEvent(nfb, nil)
        clCNFB_SetOnErrorEvent(nfb, nil)
        if NfbBridge.activeBridge === self {
            NfbBridge.activeBridge = nil
        }
    }
}

File: android/src/main/kotlin/com/neiry/neiry_kit/NfbBridge.kt
package com.neiry.neiry_kit

import io.flutter.plugin.common.EventChannel

/**
 * Kotlin bridge for the `clCNFB` classifier. Owns the native NFB handle and
 * exposes factory methods (plain and calibrated) plus stream handlers for
 * `nfbState` and `nfbError` event channels.
 *
 * Mirrors [NfbBridge.swift] in behavior.
 */
class NfbBridge(private val nativeBridge: NativeBridge) {

    private var handle: Long = 0L

    // ── Stream handlers ───────────────────────────────────────────────────────

    inner class NfbStreamHandler(
        private val setSink: (EventChannel.EventSink?) -> Unit
    ) : EventChannel.StreamHandler {
        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
            setSink(events)
        }
        override fun onCancel(arguments: Any?) {
            setSink(null)
        }
    }

    val nfbStateHandler = NfbStreamHandler { nativeBridge.nativeSetNfbStateSink(it) }
    val nfbErrorHandler = NfbStreamHandler { nativeBridge.nativeSetNfbErrorSink(it) }

    /** Returns all (channelId, handler) pairs for EventChannel registration in the plugin. */
    fun allStreamHandlers(): List<Pair<String, EventChannel.StreamHandler>> = listOf(
        "neiry_kit/events/nfbState" to nfbStateHandler,
        "neiry_kit/events/nfbError" to nfbErrorHandler,
    )

    // ── Factory methods ───────────────────────────────────────────────────────

    /** Creates an uncalibrated NFB classifier for the given device handle. */
    fun create(deviceHandle: Long) {
        if (handle != 0L) dispose()
        try {
            handle = nativeBridge.nativeCreateNfb(deviceHandle)
        } catch (e: RuntimeException) {
            throw parseSdkError(e.message ?: "255|Unknown error")
        }
    }

    /**
     * Creates a calibrated NFB classifier for the given device handle.
     * If [calibrationData] is non-null, its fields are imported into the
     * calibrator before creating the classifier.
     */
    fun createCalibrated(deviceHandle: Long, calibrationData: Map<String, Any>?) {
        if (handle != 0L) dispose()
        val hasData = calibrationData != null
        val ts                              = (calibrationData?.get("ts") as? Number)?.toLong()    ?: 0L
        val failReason                      = (calibrationData?.get("failReason") as? Number)?.toInt()  ?: 0
        val individualFrequency             = (calibrationData?.get("individualFrequency") as? Number)?.toFloat()             ?: 0f
        val individualPeakFrequency         = (calibrationData?.get("individualPeakFrequency") as? Number)?.toFloat()         ?: 0f
        val individualPeakFrequencyPower    = (calibrationData?.get("individualPeakFrequencyPower") as? Number)?.toFloat()    ?: 0f
        val individualPeakFrequencySuppression = (calibrationData?.get("individualPeakFrequencySuppression") as? Number)?.toFloat() ?: 0f
        val individualBandwidth             = (calibrationData?.get("individualBandwidth") as? Number)?.toFloat()             ?: 0f
        val individualNormalizedPower       = (calibrationData?.get("individualNormalizedPower") as? Number)?.toFloat()       ?: 0f
        val lowerFrequency                  = (calibrationData?.get("lowerFrequency") as? Number)?.toFloat()                  ?: 0f
        val upperFrequency                  = (calibrationData?.get("upperFrequency") as? Number)?.toFloat()                  ?: 0f

        try {
            handle = nativeBridge.nativeCreateNfbCalibrated(
                deviceHandle,
                hasData,
                ts,
                failReason,
                individualFrequency,
                individualPeakFrequency,
                individualPeakFrequencyPower,
                individualPeakFrequencySuppression,
                individualBandwidth,
                individualNormalizedPower,
                lowerFrequency,
                upperFrequency,
            )
        } catch (e: RuntimeException) {
            throw parseSdkError(e.message ?: "255|Unknown error")
        }
    }

    /** Unregisters callbacks and clears the NFB handle. */
    fun dispose() {
        if (handle != 0L) {
            nativeBridge.nativeDisposeNfb(handle)
            handle = 0L
        }
    }
}
